mlir/test/Catalyst/DetensorizeSCFTest.mlir

```diff
@@ -0,0 +1,254 @@
+    %2 = arith.mulf %extracted_1, %extracted_2 : f64
+    %from_elements = tensor.from_elements %2 : tensor<f64>
+    return %from_elements : tensor<f64>
+  }
+}
+
+// -----
+
+// CHECK-LABEL: @test2
+module @test2 {
+  // CHECK-LABEL: @test_for_loop
+  func.func public @test_for_loop(%arg0: tensor<f64>, %arg1: tensor<f64>) -> tensor<f64> attributes {llvm.emit_c_interface} {
+  // CHECK:     [[FOR_RESULT:.+]] = scf.for {{.*}} iter_args(%[[ARG:.+]] = {{.*}}) -> (f64) {
+  // CHECK-NOT:   tensor<
+  // CHECK:       tensor.from_elements %[[ARG]] : tensor<f64>
+  // CHECK:       [[IF_RESULT:.+]] = scf.if {{.*}} -> (f64)
+  // CHECK:         scf.yield {{.*}} : f64
+  // CHECK-NOT:     tensor<
+  // CHECK:       else
+  // CHECK:         scf.yield {{.*}} : f64
+  // CHECK-NOT:     tensor<
+  // CHECK:       }
+  // CHECK:       scf.yield {{.*}} : f64
+  // CHECK-NOT:     tensor<
+  // CHECK:     }
+  // CHECK:     arith.mulf {{.*}} [[FOR_RESULT:.+]] : f64
+    %cst = arith.constant 0.000000e+00 : f64
+    %cst_0 = arith.constant 1.000000e+00 : f64
+    %cst_1 = arith.constant dense<2> : tensor<i64>
+    %c0 = arith.constant 0 : index
+    %c10 = arith.constant 10 : index
+    %c1 = arith.constant 1 : index
+    %extracted = tensor.extract %arg0[] : tensor<f64>
+    %extracted_2 = tensor.extract %arg1[] : tensor<f64>
+    %0 = arith.addf %extracted, %extracted_2 : f64
+    %from_elements = tensor.from_elements %0 : tensor<f64>
+    %1 = scf.for %arg2 = %c0 to %c10 step %c1 iter_args(%arg3 = %from_elements) -> (tensor<f64>) {
+      %extracted_5 = tensor.extract %arg3[] : tensor<f64>
+      %3 = func.call @remainder(%arg3, %cst_1) : (tensor<f64>, tensor<i64>) -> tensor<f64>
+      %extracted_6 = tensor.extract %3[] : tensor<f64>
+      %4 = arith.cmpf une, %extracted_6, %cst : f64
+      %5 = scf.if %4 -> (f64) {
+        %6 = arith.subf %extracted_5, %extracted_5 : f64
+        scf.yield %6 : f64
+      } else {
+        %6 = arith.addf %extracted_5, %cst_0 : f64
+        scf.yield %6 : f64
+      }
+      %from_elements_7 = tensor.from_elements %5 : tensor<f64>
+      scf.yield %from_elements_7 : tensor<f64>
+    }
+    %extracted_3 = tensor.extract %1[] : tensor<f64>
+    %2 = arith.mulf %extracted, %extracted_3 : f64
+    %from_elements_4 = tensor.from_elements %2 : tensor<f64>
+    return %from_elements_4 : tensor<f64>
+  }
+  func.func private @remainder(%arg0: tensor<f64> {mhlo.layout_mode = "default"}, %arg1: tensor<i64> {mhlo.layout_mode = "default"}) -> (tensor<f64> {mhlo.layout_mode = "default"}) attributes {llvm.linkage = #llvm.linkage<internal>} {
+    %cst = arith.constant 0.000000e+00 : f64
+    %extracted = tensor.extract %arg1[] : tensor<i64>
+    %extracted_0 = tensor.extract %arg0[] : tensor<f64>
+    %0 = arith.sitofp %extracted : i64 to f64
+    %1 = arith.remf %extracted_0, %0 : f64
+    %from_elements = tensor.from_elements %1 : tensor<f64>
+    return %from_elements : tensor<f64>
+  }
```

Comment:
Looking into it a bit, we need this patch at least to be able to use the test dialect in our tests:

```diff
diff --git a/mlir/tools/quantum-opt/CMakeLists.txt b/mlir/tools/quantum-opt/CMakeLists.txt
index bc289a2c4..826778f17 100644
--- a/mlir/tools/quantum-opt/CMakeLists.txt
+++ b/mlir/tools/quantum-opt/CMakeLists.txt
@@ -13,6 +13,7 @@ set(LIBS
     MLIRGradient
     gradient-transforms
     MLIRMitigation
+    MLIRTestDialect
     mitigation-transforms
     MhloRegisterDialects
     StablehloRegister
diff --git a/mlir/tools/quantum-opt/quantum-opt.cpp b/mlir/tools/quantum-opt/quantum-opt.cpp
index b2c46933d..009a8959d 100644
--- a/mlir/tools/quantum-opt/quantum-opt.cpp
+++ b/mlir/tools/quantum-opt/quantum-opt.cpp
@@ -33,13 +33,20 @@
 #include "Quantum/IR/QuantumDialect.h"
 #include "Quantum/Transforms/Passes.h"
 
+namespace test {
+void registerTestDialect(mlir::DialectRegistry &);
+void registerTestDynDialect(mlir::DialectRegistry &);
+void registerTestTilingInterfaceTransformDialectExtension(mlir::DialectRegistry &);
+void registerTestTransformDialectExtension(mlir::DialectRegistry &);
+} // namespace test
+
 int main(int argc, char **argv)
 {
     mlir::registerAllPasses();
     catalyst::registerAllCatalystPasses();
     mlir::mhlo::registerAllMhloPasses();
-
     mlir::DialectRegistry registry;
+    ::test::registerTestDialect(registry);
     mlir::registerAllDialects(registry);
     mlir::mhlo::registerAllMhloDialects(registry);
     mlir::stablehlo::registerAllDialects(registry);
```

Reply:
Thanks for the suggestion, that is a nice idea :) What I sometimes do is use function arguments as a way to obtain arbitrary values easily, I might not add the test dialect in this PR, but I'll try and simplify the tests.

Reply:
Yeah no worries. I added the test dialect in a PR just now. I also use function arguments, but I think in some use cases, the test dialect is nicer (because we don't even need to write a function for testing)

Reply:
4a1dc28a31d6b9562afdc1df423ecf00c0ad8f37